File: src/backend/app/models/git.py
"""Git schemas (doc 04 §2.2, §13).

The working tree is the source of truth — nothing here is persisted. ``summary``
is the human roll-up the top-bar badge renders; it is built by ``GitService`` so
that every producer of a :class:`GitStatus` (the request path and the debounced
git-status worker alike) phrases it identically.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

GitFileStatus = Literal["modified", "added", "deleted", "untracked", "renamed"]


class GitFile(BaseModel):
    path: str
    status: GitFileStatus
    staged: bool


class GitStatus(BaseModel):
    dirty: bool
    files: list[GitFile] = []
    ahead: int = 0
    behind: int = 0
    hasRemote: bool = False
    branch: str = ""
    summary: str = "all-changes-synced"


class CommitInfo(BaseModel):
    hash: str
    message: str
    timestamp: str


class GitDiff(BaseModel):
    path: str
    diff: str = ""
    binary: bool = False


class SuggestedMessage(BaseModel):
    message: str
    # True when no utility model was configured and the message came from file
    # stats instead — the UI notes the provenance rather than passing it off.
    fromStats: bool = False


class RemoteResult(BaseModel):
    ok: bool = True
    summary: str = ""


# ---- request bodies ---------------------------------------------------------


class StageRequest(BaseModel):
    paths: list[str] | None = None
    all: bool | None = None

    model_config = {"extra": "forbid"}


class CommitRequest(BaseModel):
    message: str

    model_config = {"extra": "forbid"}

File: src/frontend/src/api/git.ts
import { apiGet, apiSend } from "./client";

export type GitFileStatus = "modified" | "added" | "deleted" | "untracked" | "renamed";

export interface GitFile {
  path: string;
  status: GitFileStatus;
  staged: boolean;
}

export interface GitStatus {
  dirty: boolean;
  files: GitFile[];
  ahead: number;
  behind: number;
  hasRemote: boolean;
  /** Current branch; short sha if HEAD is detached. */
  branch: string;
  /** Human roll-up for the badge: "all-changes-synced" | "7-new, 1-updated". */
  summary: string;
}

export interface CommitInfo {
  hash: string;
  message: string;
  timestamp: string;
}

export interface GitDiff {
  path: string;
  diff: string;
  binary: boolean;
}

export interface SuggestedMessage {
  message: string;
  /** True when no utility model was configured and the text came from file stats. */
  fromStats: boolean;
}

export interface RemoteResult {
  ok: boolean;
  summary: string;
}

export const getGitStatus = (bookId: string) => apiGet<GitStatus>(`/books/${bookId}/git/status`);

export const stageFiles = (bookId: string, body: { paths?: string[]; all?: boolean }) =>
  apiSend<GitStatus>("POST", `/books/${bookId}/git/stage`, body);

export const unstageFiles = (bookId: string, body: { paths?: string[]; all?: boolean }) =>
  apiSend<GitStatus>("POST", `/books/${bookId}/git/unstage`, body);

export const discardFiles = (bookId: string, body: { paths?: string[]; all?: boolean }) =>
  apiSend<GitStatus>("POST", `/books/${bookId}/git/discard`, body);

export const getGitDiff = (bookId: string, path: string) =>
  apiGet<GitDiff>(`/books/${bookId}/git/diff?path=${encodeURIComponent(path)}`);

export const suggestCommitMessage = (bookId: string) =>
  apiSend<SuggestedMessage>("POST", `/books/${bookId}/git/suggest-message`);

export const commitStaged = (bookId: string, message: string) =>
  apiSend<CommitInfo>("POST", `/books/${bookId}/git/commit`, { message });

export const pushRemote = (bookId: string) => apiSend<RemoteResult>("POST", `/books/${bookId}/git/push`);

export const pullRemote = (bookId: string) => apiSend<RemoteResult>("POST", `/books/${bookId}/git/pull`);

export const getGitLog = (bookId: string, limit = 20) =>
  apiGet<CommitInfo[]>(`/books/${bookId}/git/log?limit=${limit}`);
